lib/src/subscription.dart

```diff
@@ -0,0 +1,284 @@
+  /// subscriptions, use [waitForSynchronization].
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action);
+
+  /// Gets the version of the subscription set.
+  int get version => realmCore.subscriptionSetGetVersion(this);
+
+  /// Gets the state of the subscription set.
+  SubscriptionSetState get state {
+    final state = realmCore.subscriptionSetGetState(this);
+    switch (state) {
+      case SubscriptionSetState._uncommitted:
+      case SubscriptionSetState._bootstrapping:
+        return SubscriptionSetState.pending;
+      default:
+        return state;  
+    }
+  }
+}
+
+extension SubscriptionSetInternal on SubscriptionSet {
+  Realm get realm => _realm;
+  SubscriptionSetHandle get handle => _handle;
+
+  static SubscriptionSet create(Realm realm, SubscriptionSetHandle handle) => _ImmutableSubscriptionSet._(realm, handle);
+}
+
+class _ImmutableSubscriptionSet extends SubscriptionSet {
+  _ImmutableSubscriptionSet._(Realm realm, SubscriptionSetHandle handle) : super._(realm, handle);
+
+  @override
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action) {
+    final mutableSubscriptions = MutableSubscriptionSet._(realm, realmCore.subscriptionSetMakeMutable(this));
+    try {
+      action(mutableSubscriptions);
+      _handle = realmCore.subscriptionSetCommit(mutableSubscriptions);
+    } finally {
+      // Release as early as possible, as we cannot start new update, until this is released!
+      mutableSubscriptions._handle.release();
+    }
+  }
+}
+
+/// A mutable view to a [SubscriptionSet]. Obtained by calling [SubscriptionSet.update].
+/// {@category Sync}
+class MutableSubscriptionSet extends SubscriptionSet {
+  final MutableSubscriptionSetHandle _handle;
+
+  MutableSubscriptionSet._(Realm realm, this._handle) : super._(realm, _handle);
+
+  @override
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action) {
+    action(this);
+  }
+
+  /// Adds a [query] to the set of active subscriptions.
+  ///
+  /// The query will be joined via an OR statement with any existing queries for the same type.
+  ///
+  /// If a [name] is given, then this will be used to match with any existing query,
+  /// otherwise the [query] itself is used for matching.
+  ///
+  /// If [update] is specified to [true], then any existing query will be replaced.
+  /// Otherwise a [RealmException] is thrown, in case of duplicates.
+  /// {@category Sync}
+  Subscription add<T extends RealmObject>(RealmResults<T> query, {String? name, bool update = false}) {
+    return Subscription._(realmCore.insertOrAssignSubscription(this, query, name, update));
+  }
+
+  /// Remove the [subscription] from the set, if it exists.
+  bool remove(Subscription subscription) {
+    return realmCore.eraseSubscriptionById(this, subscription);
+  }
+
+  /// Remove the [query] from the set, if it exists.
+  bool removeByQuery<T extends RealmObject>(RealmResults<T> query) {
```

Comment:
I disagree - the notion is that we're subscribing to a query, not a result. This may be going too deep into language semantics, but "results" is a materialized query and has implications that you're subscribing to whatever you're currently seeing in this results collection, which is not the case.

Reply:
I don't like `removeByResults` here either. To be honest I would prefer having `RealmQuery` be first-class, and use it on this method. But, given we are not doing that, I think referring to it as a query is preferable.